### ISSUE TITLE
onChoosePath(String path, File pathFile) cannot have anything blocking the UI

### ISSUE THREAD

**Issue by zipswich** [hedzr/android-file-chooser#46]
For example, if you cannot use AlertDialog.  It would not show, and the app would hang.
This is not a big deal because one can always start a new thread, then grab the UI thread to remedy this.
I thought it would be nice to have the ChooserDialog closed upon entering onChoosePath(String path, File pathFile)  because it is closed anyway after this void method ends.

**Comment by Guiorgy** [comment]
What if I am suposed to choose several paths? Those kind of things are best left for the developer using to implement the exact way they want it no?

**Comment by hedzr** [comment]
You might be asking an impossible mission. :)
We can't really **dismiss** the dialog till the main ui thread looped (by Dialog#sendDismissMessage asynchronously). So `dismiss()` couldn't give you running control point (eg. cpu time), AlertDialog will yield control until its onPostiveListener ends.

Another side, as @Guiorgy just said, perhaps that's rights of developers.

**Comment by zipswich** [comment]
I am sorry for my ignorance.  I do not know how choosing several paths is related to this.

Is it possible to do something like the following in OnClickListener:
```
  dialog.dismiss();
  new Thread( ()->{
      _result.onChoosePath(_currentDir.getAbsolutePath(), _currentDir);
       ...
   }
  ).start();
```

This will close the dialog first before sending the results to the caller with a non UI thread.

**Comment by hedzr** [comment]
@zipswich based on your codes, how about this:
```java
            .withChosenListener((path, pathFile) -> {
                Toast.makeText(ctx, "FILE: " + path, Toast.LENGTH_SHORT).show(); //**not happy**

                _path = path; //**can live**
                _tv.setText(_path); //**not happy**
                //_iv.setImageURI(Uri.fromFile(pathFile)); //**not happy**
                _iv.setImageBitmap(ImageUtil.decodeFile(pathFile)); //**not happy**
            })
```
they would be not happy.

**Comment by hedzr** [comment]
wow, wrong issue number, i'm sorry. i'm busy to go to work. that's a fix for #47.

**Comment by zipswich** [comment]
I am not sure if I was confusing you.  I meant the code in the library, not the code of using this library. I was trying to suggest code to make the diaglog disappear before calling the listener.

**Comment by hedzr** [comment]
Actually, i figure that somebody may not take benifits while we apply the new codes like yours:
```java
new Thread( ()->{
      _result.onChoosePath(_currentDir.getAbsolutePath(), _currentDir);
       ...
   }
  ).start();
```
And `_result.onChoosePath` will be put in a thread different from main ui, so developer couldn't extract `path` to another ui control. That's why this code not happy:
```java
            .withChosenListener((path, pathFile) -> {
                _tv.setText(_path); // _tv=findViewById(R.id.textView)
            })
```

Did I make a mistake?
> my noon, my reply

**Comment by zipswich** [comment]
Understood.  
In this case,  runOnUiThread() could be used to return the results:
```
  dialog.dismiss();
  new Thread( ()->{
        activity.runOnUiThead(result.onChoosePath(_currentDir.getAbsolutePath(), _currentDir));
   }
  ).start();
```

This will achieve closing the dialog before returning the results on the UI thread.

**Comment by hedzr** [comment]
It's quite a right way to solve the matter which not in the same thread. but new question is: why all of us need two nesting calls here?
so, let's back to the first problem. you said:
> if you cannot use AlertDialog. It would not show, and the app would hang

the scene i can imagine is you are picking file from a non-ui thread, isn't it?
so the better thing should be runOnUiThread(newPicker()) at developer side, rather than making double nesting runnable calls from library to developer.

it's a developer's choice but library's, my sincerely.

and, when we changed the behavior of `onChoosePath()`, such as wrapped within a single runnable, the existing legacy codes would had been modified.
that's another reason why we should talk about this issue more, we need to strike a balance.

**Comment by zipswich** [comment]
Using Thread > runOnUiThread is how I do it in the code to deal with this.  It works fine.  This is definitely not a show stopper problem. Once a library user realizes this, he can do the same thing.

I would think the expectation (at least it was my expectation) of library users is that once the results are returned, the job of ChooserDialog is completely finished, he should not worry about anything related to it.  Right now, the users need to keep in mind that ChooserDialog is not closed yet, and their code needs to return the UI thread as soon as possible. 

In other words, when OK or Cancel button is clicked, the dialog is expected to close unconditionally regardless of what the library user's code does.   Whatever happens afterwards is the full responsibility of the library user.

**Comment by hedzr** [comment]
Using RxAndroid.
It's not good to start a thread in place while `path` picked. Moving the long time action into an observer, putting the `path` into an observable field.
observers' threads are reusable. but the thread in place is unmanageable.

The library user needs not to keep in mind **when** and how the dialog is closed. He concerns to try to finding the best way to split ui and non-ui codes, to split the asynchronous actions, to reuse those repeatable threads, thread-pool or executor. These can make ui responsive effectively.

**Comment by zipswich** [comment]
Fire enough. I just have a very simple blocking confirmation dialog which, of course, can be improved simply to accommodate this.
